Environmental renderers

Waterpicker opened the pull request:
I don't know how to write this in technical babble so I'll be blunt about it.

SkyProperties are hardcoded so need a way to build and register custom ones.

Custom Sky/Cloud/Weather rendering already requires mixins so I just make hooks for them.

An example of why having custom Sky Rendering hooked this way is a mod I personally work on, Dimensional Doors.

In the new datapack dungeon system, pockets can have various attributes. Among them is the ability to specify the sky rendering of other dimensions as if they were self contained demiplanes of D&D. 

Another example where this PR would be useful is the fabric port of Twilight Forest that current has a completely custom SkyProperties class and custom Sky/Weather renderers.

```java
DimensionRenderingRegistry.setSkyRenderer(World.OVERWORLD, (world, matrices, tickDelta) -> {
    // Custom sky Rendering code
});
```

Waterpicker committed:
Environmental Renderers

Waterpicker committed:
Environmental Renderers

Waterpicker committed:
Remove defunt hideClouds and hideWeather

Waterpicker committed:
Remove defunt hideClouds and hideWeather

Waterpicker committed:
Pushed missing License

Waterpicker committed:
Pushed missing License

Waterpicker committed:
Push checkstyle fixes

Waterpicker committed:
Push checkstyle fixes

frqnny commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,137 @@
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.util.math.Vec3d;
+
+public final class FabricSkyPropertyBuilder {
```
Some javadocs here would be great, specially since this is a public API.

Waterpicker replied:
I 

> 
> 
> Some javadocs here would be great, specially since this is a public API.

I tried to make a few javadocs. Could use some help with wording with some. Documentation isn't my strong suit. I had help with the other api's docs.

frqnny commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinWorldRenderer.java:
```diff
@@ -166,8 +167,44 @@ private void afterRender(CallbackInfo ci) {
+	private void renderWeather(LightmapTextureManager manager, float tickDelta, double x, double y, double z, CallbackInfo info) {
+		EnvironmentRenderers.WeatherRenderer renderer = null;
+
+		if (this.client.world != null) renderer = EnvironmentRenderers.getWeatherRenderer(world.getRegistryKey());
```
Checkstyle wants you to always use brackets with if statements. Just giving you a heads up (I might be wrong, though)

frqnny commented on fabric-rendering-v1/src/testmod/java/net/fabricmc/fabric/test/rendering/client/EnvironementalRenderingTest.java:
```diff
@@ -0,0 +1,91 @@
+			for (int i = 0; i < 6; ++i) {
+				matrices.push();
+
+				if (i == 1) {
```
Couldn't you simplify this code with a switch statement?

frqnny reviewed (commented):
I've Never touched this general area of the game, just thought to leave some thoughts here and there.

AlphaMode committed:
remove unused

AlphaMode committed:
remove unused

Waterpicker committed:
Merge pull request #1 from AlphaMode/patch-1

remove unused

Waterpicker committed:
Unwrapped the test mod sky rendering.

Waterpicker committed:
Unwrapped the test mod sky rendering.

Waterpicker committed:
Merge remote-tracking branch 'Waterpicker/environmental-renderers' into environmental-renderers

Waterpicker committed:
Added a first attempt at javadoc.

Waterpicker committed:
Added a first attempt at javadoc.

Waterpicker committed:
Forgot a blank line

Waterpicker committed:
Forgot a blank line

liach commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/EnvironmentRenderers.java:
```diff
@@ -0,0 +1,126 @@
+	private static final Map<RegistryKey<World>, SkyRenderer> SKY_RENDERERS = new IdentityHashMap<>();
+	@Environment(EnvType.CLIENT)
+	private static final Map<RegistryKey<World>, CloudRenderer> CLOUD_RENDERERS = new HashMap<>();
+	@Environment(EnvType.CLIENT)
```
Please move all these environment annotations onto the classes itself than leaving them on methods and fields. Putting them on fields is definitely wrong as that will break the initializer code at runtime.

liach commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,195 @@
+	private boolean alternateSkyColor = true;
+
+	private BiFunction<Vec3d, Float, Vec3d> adjustFogColor = (color, sunHeight) -> color.multiply((sunHeight * 0.94F + 0.06F), sunHeight * 0.94F + 0.06F, (sunHeight * 0.91F + 0.09F));
+	private BiPredicate<Integer, Integer> useThickFog = (camX, camY) -> false;
```
I recommend dedicated functional interfaces that have their parameters named for clarity.

Waterpicker committed:
Moved Environment references out of class

Waterpicker committed:
Moved Environment references out of class

Waterpicker committed:
Moved Lamda functions for fog override, thick fog usage and fog color adjustment to their own interfaces and gave them proper javadoc.

Waterpicker committed:
Moved Lamda functions for fog override, thick fog usage and fog color adjustment to their own interfaces and gave them proper javadoc.

liach commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,240 @@
+		return new FabricSkyproperties(cloudsHeight, alternateSkyColor, skyType, brightenLighting, darkened, fogColorOverride, adjustFogColor, useThickFog);
+	}
+
+	public static class FabricSkyproperties extends SkyProperties {
```
Is this class supposed to be visible to users? If not making it private or package private is better, so we can change it freely like it's an implementation detail.

Waterpicker replied:
Its suppose to implementation.

liach replied:
Then I suggest
```suggestion
	private static class FabricSkyproperties extends SkyProperties {
```
so modders cannot use it in instanceof checks, call its methods etc. explicitly, and we can change this class as part of the implementaion however we want to.

Waterpicker replied:
Done.

liach commented:
Thanks, now this api looks much cleaner!

Waterpicker committed:
Made FabricSkyProperties private

Waterpicker committed:
Made FabricSkyProperties private

Juuxel commented on fabric-rendering-v1/src/testmod/java/net/fabricmc/fabric/test/rendering/client/EnvironementalRenderingTest.java:
```diff
@@ -0,0 +1,94 @@
+import net.fabricmc.fabric.api.client.rendering.v1.EnvironmentRenderers;
+import net.fabricmc.fabric.api.client.rendering.v1.FabricSkyPropertyBuilder;
+
+public class EnvironementalRenderingTest implements ClientModInitializer {
```
Typo here: should be "Environmental" (also in fabric.mod.json since this is an entrypoint)

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,240 @@
+	 * Says if lighting should be darker.
+	 *
+	 * @param darkened should lighting be darker
+	 * @return a reference to the FabricItemGroupBuilder
```
Item group builder? This (and the others) can just be `@return this builder` or something IMO

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/EnvironmentRenderers.java:
```diff
@@ -0,0 +1,124 @@
+import net.fabricmc.fabric.mixin.client.rendering.SkyPropertiesAccessor;
+
+@Environment(EnvType.CLIENT)
+public final class EnvironmentRenderers {
```
This could do with a short class javadoc comment explaining what it's for (for example, what "environment renderers" are)

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,240 @@
+
+/**
+ * Builder used to create {@link net.minecraft.client.render.SkyProperties} instances which
+ * govern various things about how a dimension renders including where thick fog should be used, what type of should be rendered, et ect.
```
> what type of should be rendered

1\) Type of what? I guess this is missing a word. 2) Also:

```suggestion
 * govern various things about how a dimension renders including where thick fog should be used, what type of should be rendered, etc.
```

Juuxel reviewed (changes requested):
The api looks good, just some minor comments

NinjaPhenix commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,240 @@
+	@FunctionalInterface
+	public interface UseThickFog {
+		/**
+		 * Used to determine if thick fog should be used current xy coordinates of teh camera.
```
`of the camera.` rather than `teh`.

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/EnvironmentRenderers.java:
```diff
@@ -0,0 +1,124 @@
+	 * @param key A RegistryKey for your Dimension Type
+	 * @param renderer A {@link SkyRenderer} implementation
+	 */
+	@Environment(EnvType.CLIENT)
```
Note that annotating every single member is not necessary, just annotating the class is enough.

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/EnvironmentRenderers.java:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	@Environment(EnvType.CLIENT)
+	public static SkyRenderer getSkyRenderer(RegistryKey<World> key) {
```
I'd add `@Nullable`. And same thing for the other two below.

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,240 @@
+	 * @return a reference to the FabricItemGroupBuilder
+	 */
+	public FabricSkyPropertyBuilder adjustFogColor(AdjustFogColor adjustFogColor) {
+		this.adjustFogColor = adjustFogColor;
```
`Objects.requireNonNull`?

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,240 @@
+		return new FabricSkyproperties(cloudsHeight, alternateSkyColor, skyType, brightenLighting, darkened, fogColorOverride, adjustFogColor, useThickFog);
+	}
+
+	private static class FabricSkyproperties extends SkyProperties {
```
```suggestion
	private static class FabricSkyProperties extends SkyProperties {
```

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,240 @@
+	}
+
+	private static class FabricSkyproperties extends SkyProperties {
+		private final float[] color = new float[4];
```
I can't find where this is used.

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,240 @@
+		}
+
+		@Override
+		public float[] getFogColorOverride(float skyAngle, float tickDelta) {
```
`@Nullable`.

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinWorldRenderer.java:
```diff
@@ -166,8 +167,44 @@ private void afterRender(CallbackInfo ci) {
 	}
 
-	@Inject(method = "Lnet/minecraft/client/render/WorldRenderer;reload()V", at = @At("HEAD"))
+	@Inject(method = "reload()V", at = @At("HEAD"))
```
Why? Wouldn't this undo #1701?

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinWorldRenderer.java:
```diff
@@ -41,17 +41,18 @@
 
 @Mixin(WorldRenderer.class)
 public abstract class MixinWorldRenderer {
+	@Unique private final WorldRenderContextImpl context = new WorldRenderContextImpl();
```
Why move this around? No big deal, but I'd avoid this (seemingly?) unnecessary change.

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,240 @@
+ * govern various things about how a dimension renders including where thick fog should be used, what type of should be rendered, et ect.
+ */
+@Environment(EnvType.CLIENT)
+public final class FabricSkyPropertyBuilder {
```
From the looks of it, subclassing `SkyProperties` would achieve the same thing. Why do we need a builder?

shartte commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinWorldRenderer.java:
```diff
@@ -166,8 +167,44 @@ private void afterRender(CallbackInfo ci) {
+	private void renderSky(MatrixStack matrices, Matrix4f matrix4f, float tickDelta, Runnable runnable, CallbackInfo info) {
+		EnvironmentRenderers.SkyRenderer renderer = null;
+
+		if (this.client.world != null) renderer = EnvironmentRenderers.getSkyRenderer(world.getRegistryKey());
```
Formatting.
How does this pass checkstyle???

Juuxel replied:
It allows single-line `if`s and such:
https://github.com/FabricMC/fabric/blob/46582230fb580d4c1f71e4b0737df27417ec9cb1/checkstyle.xml#L91-L94

shartte commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/EnvironmentRenderers.java:
```diff
@@ -0,0 +1,124 @@
+	private static final Map<RegistryKey<World>, WeatherRenderer> WEATHER_RENDERERS = new HashMap<>();
+
+	/**
+	 * Registers a custom sky renderer for a DimensionType.
+	 *
+	 * @param key A RegistryKey for your Dimension Type
+	 * @param renderer A {@link SkyRenderer} implementation
```
```suggestion
	 * Sets a custom sky renderer for a world.
	 * <p>
	 * This overrides Vanilla's sky rendering.
	 *
	 * @param key The registry key of the world.
	 * @param renderer The {@link SkyRenderer} to use for the world.
```

shartte replied:
And similar comments for the other methods, I guess.

shartte commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/EnvironmentRenderers.java:
```diff
@@ -0,0 +1,124 @@
+
+	@FunctionalInterface
+	public interface SkyRenderer {
+		void render(MinecraftClient world, MatrixStack matrices, float tickDelta);
```
```suggestion
		void render(MinecraftClient client, MatrixStack matrices, float tickDelta);
```

shartte commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,240 @@
+
+		if (g >= -0.4F && g <= 0.4F) {
+			float i = (g - -0.0F) / 0.4F * 0.5F + 0.5F;
+			float j = 1.0F - (1.0F - MathHelper.sin(i * 3.1415927F)) * 0.99F;
```
```suggestion
			float j = 1.0F - (1.0F - MathHelper.sin(i * MathHelper.PI)) * 0.99F;
```

shartte commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,240 @@
+@Environment(EnvType.CLIENT)
+public final class FabricSkyPropertyBuilder {
+	private FogColorOverride fogColorOverride = (color, skyAngle, tickDelta) -> {
+		float g = MathHelper.cos(skyAngle * 6.2831855F) - 0.0F;
```
```suggestion
		float g = MathHelper.cos(skyAngle * MathHelper.TWO_PI);
```

shartte commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/FabricSkyPropertyBuilder.java:
```diff
@@ -0,0 +1,240 @@
+ */
+@Environment(EnvType.CLIENT)
+public final class FabricSkyPropertyBuilder {
+	private FogColorOverride fogColorOverride = (color, skyAngle, tickDelta) -> {
```
Can you remove the default value here completely and instead of copying the vanilla code here call the super-method?

Waterpicker commented:
FabricSkyPropertiesBuilder is being removed next PR due to conversation with @Technici4n on fabricord.

Waterpicker committed:
PURGE THE H...I mean FabricSkyPropertyBuilder and FabricSkyProperties.

- Reverted unintentional mixin changes
- Fix various typo errors that were pointed out.
- Added missing javadocs and added the ability to override.

Waterpicker committed:
PURGE THE H...I mean FabricSkyPropertyBuilder and FabricSkyProperties.

- Reverted unintentional mixin changes
- Fix various typo errors that were pointed out.
- Added missing javadocs and added the ability to override.

Waterpicker committed:
Forgot some pluggin in some overrides.

Waterpicker committed:
Forgot some pluggin in some overrides.

Waterpicker committed:
Added missing javadoc suggestion

Waterpicker committed:
Added missing javadoc suggestion

Technici4n reviewed (approved):
Looks fantastic.

modmuss50 commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinWorldRenderer.java:
```diff
@@ -170,4 +171,40 @@ private void afterRender(CallbackInfo ci) {
+		if (this.client.world != null) renderer = EnvironmentRenderers.getCloudRenderer(world.getRegistryKey());
+
+		if (renderer != null) {
+			renderer.render(this.client, matrices, matrix4f, tickDelta, cameraX, cameraY, cameraZ);
```
Would it be useful if the renderer returned a boolean to decide if the vanilla logic should still be used? Would make it useful to render stuff alongside the vanilla stuff?

Waterpicker replied:
Hmmmmmmmm. I just reread your stuff. That could be interesting. I'm however gonna add you don't have to get an API 100%perfect the first time. Iteration is a thing after all.

modmuss50 commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/EnvironmentRenderers.java:
```diff
@@ -0,0 +1,163 @@
+	 * <p>This overrides Vanilla's weather rendering.
+	 * @param key A RegistryKey for your Dimension Type
+	 * @param renderer A {@link WeatherRenderer} implementation
+	 * @param override Should try to override current SkyRenderer if it exists
```
Its not ideal that only one mod can register a renderer, I guess a mod could chain the renderers by calling `get*Renderer` first? Is this something that could/should be handled better? Mod load order is none deremisitic so its technically possible for your sky to change each time to load the game?

Could be something we just accept.

Waterpicker replied:
If anyone has suggestion, feel free to propose.

Waterpicker committed:
Changed api a bit so that it fires a exception if renderer not being overridden and one already exists.

Waterpicker committed:
Changed api a bit so that it fires a exception if renderer not being overridden and one already exists.

Waterpicker committed:
Split Implemtnation and Interface

Waterpicker committed:
Split Implemtnation and Interface

Waterpicker committed:
Forgot a license

Waterpicker committed:
Forgot a license

Waterpicker committed:
Added WorldRenderContext to Renderers

Waterpicker committed:
Added WorldRenderContext to Renderers

Waterpicker committed:
Added logging

Waterpicker committed:
Added logging

shartte commented:
```diff
diff --git a/fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java b/fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java
--- fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java
+++ fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java
@@ -89,14 +89,15 @@
 	void setSkyProperty(RegistryKey<DimensionType> key, SkyProperties properties, boolean override);
 
 	/**
 	 * Registers a custom sky property for a {@link DimensionType}.
+	 * <p>
+	 * This overrides Vanilla's default {@link SkyProperties}.
 	 *
-	 *  <p>This overrides Vanilla's default {@link SkyProperties}.
 	 * @param key A {@link RegistryKey} for your {@link DimensionType}
 	 * @param properties The {@link DimensionType}'s {@link SkyProperties}
 	 */
-	void setSkyProperty(RegistryKey<DimensionType> key, SkyProperties properties);
+	void setSkyProperties(RegistryKey<DimensionType> key, SkyProperties properties);
 
 	/**
 	 * Registers a custom cloud renderer for a {@link World}.
 	 *
@@ -115,19 +116,35 @@
 	 * @param renderer A {@link CloudRenderer} implementation
 	 */
 	void setCloudRenderer(RegistryKey<World> key, CloudRenderer renderer);
 
+	/**
+	 * Gets the custom sky renderer for the given dimension.
+	 * @return Null if no custom sky renderer is registered for the dimension.
+	 */
 	@Nullable
 	SkyRenderer getSkyRenderer(RegistryKey<World> key);
 
+	/**
+	 * Gets the custom cloud renderer for the given dimension.
+	 * @return Null if no custom cloud renderer is registered for the dimension.
+	 */
 	@Nullable
 	CloudRenderer getCloudRenderer(RegistryKey<World> key);
 
+	/**
+	 * Gets the custom weather effect renderer for the given dimension.
+	 * @return Null if no custom weather effect renderer is registered for the dimension.
+	 */
 	@Nullable
 	WeatherRenderer getWeatherRenderer(RegistryKey<World> key);
 
+	/**
+	 * Gets the custom sky properties for a dimension type. Returns null if
+	 * no custom sky properties are set.
+	 */
 	@Nullable
-	SkyProperties getSkyProperty(RegistryKey<DimensionType> key);
+	SkyProperties getSkyProperties(RegistryKey<DimensionType> key);
 
 	@FunctionalInterface
 	interface SkyRenderer {
 		void render(WorldRenderContext context);

```

shartte commented:
```diff
diff --git a/fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java b/fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java
--- fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java
+++ fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java
@@ -89,14 +89,15 @@
 	void setSkyProperty(RegistryKey<DimensionType> key, SkyProperties properties, boolean override);
 
 	/**
 	 * Registers a custom sky property for a {@link DimensionType}.
+	 * <p>
+	 * This overrides Vanilla's default {@link SkyProperties}.
 	 *
-	 *  <p>This overrides Vanilla's default {@link SkyProperties}.
 	 * @param key A {@link RegistryKey} for your {@link DimensionType}
 	 * @param properties The {@link DimensionType}'s {@link SkyProperties}
 	 */
-	void setSkyProperty(RegistryKey<DimensionType> key, SkyProperties properties);
+	void setSkyProperties(RegistryKey<DimensionType> key, SkyProperties properties);
 
 	/**
 	 * Registers a custom cloud renderer for a {@link World}.
 	 *
@@ -115,19 +116,35 @@
 	 * @param renderer A {@link CloudRenderer} implementation
 	 */
 	void setCloudRenderer(RegistryKey<World> key, CloudRenderer renderer);
 
+	/**
+	 * Gets the custom sky renderer for the given dimension.
+	 * @return Null if no custom sky renderer is registered for the dimension.
+	 */
 	@Nullable
 	SkyRenderer getSkyRenderer(RegistryKey<World> key);
 
+	/**
+	 * Gets the custom cloud renderer for the given dimension.
+	 * @return Null if no custom cloud renderer is registered for the dimension.
+	 */
 	@Nullable
 	CloudRenderer getCloudRenderer(RegistryKey<World> key);
 
+	/**
+	 * Gets the custom weather effect renderer for the given dimension.
+	 * @return Null if no custom weather effect renderer is registered for the dimension.
+	 */
 	@Nullable
 	WeatherRenderer getWeatherRenderer(RegistryKey<World> key);
 
+	/**
+	 * Gets the custom sky properties for a dimension type. Returns null if
+	 * no sky properties are registered.
+	 */
 	@Nullable
-	SkyProperties getSkyProperty(RegistryKey<DimensionType> key);
+	SkyProperties getSkyProperties(RegistryKey<DimensionType> key);
 
 	@FunctionalInterface
 	interface SkyRenderer {
 		void render(WorldRenderContext context);

```

shartte commented:
```diff
diff --git a/fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java b/fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java
--- fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java
+++ fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java
@@ -89,14 +89,15 @@
 	void setSkyProperty(RegistryKey<DimensionType> key, SkyProperties properties, boolean override);
 
 	/**
 	 * Registers a custom sky property for a {@link DimensionType}.
+	 * <p>
+	 * This overrides Vanilla's default {@link SkyProperties}.
 	 *
-	 *  <p>This overrides Vanilla's default {@link SkyProperties}.
 	 * @param key A {@link RegistryKey} for your {@link DimensionType}
 	 * @param properties The {@link DimensionType}'s {@link SkyProperties}
 	 */
-	void setSkyProperty(RegistryKey<DimensionType> key, SkyProperties properties);
+	void setSkyProperties(RegistryKey<DimensionType> key, SkyProperties properties);
 
 	/**
 	 * Registers a custom cloud renderer for a {@link World}.
 	 *
@@ -115,19 +116,35 @@
 	 * @param renderer A {@link CloudRenderer} implementation
 	 */
 	void setCloudRenderer(RegistryKey<World> key, CloudRenderer renderer);
 
+	/**
+	 * Gets the custom sky renderer for the given dimension.
+	 * @return Null if no custom sky renderer is registered for the dimension.
+	 */
 	@Nullable
 	SkyRenderer getSkyRenderer(RegistryKey<World> key);
 
+	/**
+	 * Gets the custom cloud renderer for the given dimension.
+	 * @return Null if no custom cloud renderer is registered for the dimension.
+	 */
 	@Nullable
 	CloudRenderer getCloudRenderer(RegistryKey<World> key);
 
+	/**
+	 * Gets the custom weather effect renderer for the given dimension.
+	 * @return Null if no custom weather effect renderer is registered for the dimension.
+	 */
 	@Nullable
 	WeatherRenderer getWeatherRenderer(RegistryKey<World> key);
 
+	/**
+	 * Gets the custom sky properties for a dimension type. Returns null if
+	 * no sky properties are registered.
+	 */
 	@Nullable
-	SkyProperties getSkyProperty(RegistryKey<DimensionType> key);
+	SkyProperties getSkyProperties(RegistryKey<DimensionType> key);
 
 	@FunctionalInterface
 	interface SkyRenderer {
 		void render(WorldRenderContext context);

```

Waterpicker committed:
Added missing javadocs and removed the override option.

Waterpicker committed:
Added missing javadocs and removed the override option.

Waterpicker commented:
> 
> 
> ```diff
> diff --git a/fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java b/fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java
> --- fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java
> +++ fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java
> @@ -89,14 +89,15 @@
>  	void setSkyProperty(RegistryKey<DimensionType> key, SkyProperties properties, boolean override);
>  
>  	/**
>  	 * Registers a custom sky property for a {@link DimensionType}.
> +	 * <p>
> +	 * This overrides Vanilla's default {@link SkyProperties}.
>  	 *
> -	 *  <p>This overrides Vanilla's default {@link SkyProperties}.
>  	 * @param key A {@link RegistryKey} for your {@link DimensionType}
>  	 * @param properties The {@link DimensionType}'s {@link SkyProperties}
>  	 */
> -	void setSkyProperty(RegistryKey<DimensionType> key, SkyProperties properties);
> +	void setSkyProperties(RegistryKey<DimensionType> key, SkyProperties properties);
>  
>  	/**
>  	 * Registers a custom cloud renderer for a {@link World}.
>  	 *
> @@ -115,19 +116,35 @@
>  	 * @param renderer A {@link CloudRenderer} implementation
>  	 */
>  	void setCloudRenderer(RegistryKey<World> key, CloudRenderer renderer);
>  
> +	/**
> +	 * Gets the custom sky renderer for the given dimension.
> +	 * @return Null if no custom sky renderer is registered for the dimension.
> +	 */
>  	@Nullable
>  	SkyRenderer getSkyRenderer(RegistryKey<World> key);
>  
> +	/**
> +	 * Gets the custom cloud renderer for the given dimension.
> +	 * @return Null if no custom cloud renderer is registered for the dimension.
> +	 */
>  	@Nullable
>  	CloudRenderer getCloudRenderer(RegistryKey<World> key);
>  
> +	/**
> +	 * Gets the custom weather effect renderer for the given dimension.
> +	 * @return Null if no custom weather effect renderer is registered for the dimension.
> +	 */
>  	@Nullable
>  	WeatherRenderer getWeatherRenderer(RegistryKey<World> key);
>  
> +	/**
> +	 * Gets the custom sky properties for a dimension type. Returns null if
> +	 * no custom sky properties are set.
> +	 */
>  	@Nullable
> -	SkyProperties getSkyProperty(RegistryKey<DimensionType> key);
> +	SkyProperties getSkyProperties(RegistryKey<DimensionType> key);
>  
>  	@FunctionalInterface
>  	interface SkyRenderer {
>  		void render(WorldRenderContext context);
> ```

In finished it.

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,123 @@
+	 * The singleton instance of the renderer registry.
+	 * Use this instance to call the methods in this interface.
+	 */
+	DimensionRenderingRegistry INSTANCE = DimensionRenderingRegistryImpl.INSTANCE;
```
Should move away from this pattern and use static functions instead.

Waterpicker committed:
Removed INSTANCE

Waterpicker committed:
Removed INSTANCE

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,116 @@
+import net.fabricmc.fabric.mixin.client.rendering.SkyPropertiesAccessor;
+
+public final class DimensionRenderingRegistryImpl {
+	Logger logger = LogManager.getLogger("FabricDimensionRenderingRegistry");
```
Looks unused.

temporadev replied:
Might as well remove it then.
```suggestion
```

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,139 @@
+public interface DimensionRenderingRegistry {
+
+	/**
+	 * sets the custom sky renderer for a {@link World}.
```
Capital s maybe?

Juuxel replied:
```suggestion
	 * Sets the custom sky renderer for a {@link World}.
```

Technici4n reviewed (approved):
Looks great now!

Waterpicker committed:
Fixed issues

Waterpicker committed:
Fixed Issues

Waterpicker committed:
Fixed formatting issue.

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,116 @@
+		Objects.requireNonNull(key);
+		Objects.requireNonNull(renderer);
+
+		DimensionRenderingRegistry.SkyRenderer prior = SKY_RENDERERS.get(key);
```
Since the value of `prior` is not used in these, it can be just a `containsKey(key)` call instead of `get(key) != null` (applies to the other 3 too)

Juuxel commented on fabric-rendering-v1/src/testmod/java/net/fabricmc/fabric/test/rendering/client/DimensionalRenderingTest.java:
```diff
@@ -0,0 +1,88 @@
+
+	@Override
+	public void onInitializeClient() {
+		DimensionRenderingRegistry.setSkyRenderer(World.OVERWORLD, DimensionalRenderingTest::render);
```
It'd be nice to also test the other renderers, but I guess this works too.

temporadev commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	public static void setSkyProperties(RegistryKey<DimensionType> key, SkyProperties properties) {
+		Objects.requireNonNull(key);
```
Could this just be simplified with `@NotNull` in the parameters?

haykam821 replied:
`@NotNull` does not check at runtime.

temporadev replied:
Ah I see, nevermind then.

temporadev commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,116 @@
+		}
+	}
+
+	public static void setSkyProperties(RegistryKey<DimensionType> key, SkyProperties properties) {
```
```suggestion
	public static void setSkyProperties(@NotNull RegistryKey<DimensionType> key, @NotNull SkyProperties properties) {
```

temporadev commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,116 @@
+
+	public static void setSkyProperties(RegistryKey<DimensionType> key, SkyProperties properties) {
+		Objects.requireNonNull(key);
+		Objects.requireNonNull(properties);
```
```suggestion
```

temporadev commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	public static void setSkyProperties(RegistryKey<DimensionType> key, SkyProperties properties) {
+		Objects.requireNonNull(key);
```
```suggestion
```

temporadev commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	public static void setCloudRenderer(RegistryKey<World> key, CloudRenderer renderer) {
+		Objects.requireNonNull(key);
```
Same here. This could be simplified to a `@NotNull` in the parameters

temporadev reviewed (changes requested):
Looking good. I will probably use this once merged.
There are some `Objects.requireNonNull(...);` that could be simplified though.

Waterpicker committed:
Merge remote-tracking branch 'FabricMC/1.17' into environmental-renderers

Waterpicker commented:
Um quick warning. While attempting to do a switch over to a local maven copy of the new api by request of the lead of the fabric twilight forest port, discovered that SkyProperties won't register due to the fact the map holding properties returns the overworld one when null is detected. An oversight on my part and attempting to fix now.

Waterpicker committed:
completed Juuxl's suggestions and removed null check on adding skyproperties. Changed SkyProperties registering to take an Identifier. Registration turns out to be only adding new options to the "effects" entry of dimension type jsonjson.

Waterpicker commented on fabric-gametest-api-v1/build.gradle:
```diff
@@ -12,7 +12,6 @@ loom {
 			server()
 			name "Game Test"
 			vmArg "-Dfabric-api.gametest"
-			vmArg "-Dfabric-api.gametest.report-file=${project.buildDir}/junit.xml"
```
Hehe oops let me reverse this. Forgot I removed this bit locally due to issue testing on going into multplayer due to weird bugs from a version of the library that I adpated into this codebase!

Waterpicker committed:
Forgot to reenable this.

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,109 @@
+		Objects.requireNonNull(key);
+		Objects.requireNonNull(renderer);
+
+		if (WEATHER_RENDERERS.get(key) != null) {
```
```suggestion
		if (WEATHER_RENDERERS.containsKey(key)) {
```

liach replied:
hould just call putIfAbsent instead

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,109 @@
+		Objects.requireNonNull(key);
+		Objects.requireNonNull(renderer);
+
+		if (CLOUD_RENDERERS.get(key) != null) {
```
```suggestion
		if (CLOUD_RENDERERS.containsKey(key)) {
```

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,109 @@
+		Objects.requireNonNull(key);
+		Objects.requireNonNull(renderer);
+
+		DimensionRenderingRegistry.SkyRenderer prior = SKY_RENDERERS.get(key);
+
+		if (prior != null) {
```
```suggestion
		if (SKY_RENDERERS.containsKey(key)) {
```

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,138 @@
+	 * @param key        The {@link net.minecraft.util.Identifier} for the new option entry.ide
+	 * @param properties The {@link SkyProperties} option.
+	 */
+	static void setSkyProperties(Identifier key, SkyProperties properties) {
```
IMO `register*` is better for these identifier methods (compare with `Registry.register` and other FAPI registries etc)

Waterpicker committed:
Just did suggested tweaks

Waterpicker committed:
Merge remote-tracking branch 'FabricMC/1.17' into environmental-renderers

Waterpicker committed:
Merge remote-tracking branch 'FabricMC/1.17' into environmental-renderers

Waterpicker committed:
Changed SkyProperties to DimensionEffects

haykam821 commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,138 @@
+	 * Gets the custom cloud renderer for the given {@link World}.
+	 *
+	 * @param key A {@link RegistryKey} for your {@link World}
+	 * @return Null if no custom cloud renderer is registered for the dimension.
```
```suggestion
	 * @return {@code null} if no custom cloud renderer is registered for the dimension.
```

haykam821 commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,138 @@
+	/**
+	 * Gets the custom weather effect renderer for the given {@link World}.
+	 *
+	 * @return Null if no custom weather effect renderer is registered for the dimension.
```
```suggestion
	 * @return {@code null} if no custom weather effect renderer is registered for the dimension.
```

haykam821 commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinWorldRenderer.java:
```diff
@@ -170,4 +171,40 @@ private void afterRender(CallbackInfo ci) {
+		if (renderer != null) {
+			renderer.render(context);
+			info.cancel();
+		}
```
```suggestion
		if (this.client.world != null) {
			DimensionRenderingRegistry.SkyRenderer renderer = DimensionRenderingRegistry.getSkyRenderer(world.getRegistryKey());
			if (renderer != null) {
				renderer.render(context);
				info.cancel();
			}
		}
```

haykam821 commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinWorldRenderer.java:
```diff
@@ -170,4 +171,40 @@ private void afterRender(CallbackInfo ci) {
+		if (renderer != null) {
+			renderer.render(context);
+			info.cancel();
+		}
```
```suggestion
		if (this.client.world != null) {
			DimensionRenderingRegistry.CloudRenderer renderer = DimensionRenderingRegistry.getCloudRenderer(world.getRegistryKey());
			if (renderer != null) {
				renderer.render(context);
				info.cancel();
			}
		}
```

haykam821 commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinWorldRenderer.java:
```diff
@@ -170,4 +171,40 @@ private void afterRender(CallbackInfo ci) {
+		if (renderer != null) {
+			renderer.render(context);
+			info.cancel();
+		}
```
```suggestion
		if (this.client.world != null) {
			DimensionRenderingRegistry.WeatherRenderer renderer = DimensionRenderingRegistry.getWeatherRenderer(world.getRegistryKey());
			if (renderer != null) {
				renderer.render(context);
				info.cancel();
			}
		}
```

haykam821 commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,138 @@
+	/**
+	 * Sets the custom sky renderer for a {@link World}.
+	 *
+	 * <p>This overrides Vanilla's sky rendering.
```
I would document that this method throws an `IllegalStateException` if the world already has a custom sky renderer registered.

haykam821 commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,138 @@
+	/**
+	 * Registers a custom weather renderer for a {@link World}.
+	 *
+	 * <p>This overrides Vanilla's weather rendering.
```
I would document that this method throws an `IllegalStateException` if the world already has a custom weather renderer registered.

haykam821 commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,138 @@
+	/**
+	 * Registers a custom cloud renderer for a {@link World}.
+	 *
+	 * <p>This overrides Vanilla's cloud rendering.
```
I would document that this method throws an `IllegalStateException` if the world already has a custom cloud renderer registered.

haykam821 commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,138 @@
+	/**
+	 * Registers a sky property for a {@link net.minecraft.util.Identifier}.
+	 *
+	 * <p>This registers a new option for the "effects" entry of the dimension type json.
```
I would document that this method throws an `IllegalStateException` if the world already has a custom sky property registered.

apple502j commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	/**
+	 * Registers a sky property for a {@link net.minecraft.util.Identifier}.
```
This is called "dimension effects" after the yarn rename. Maybe rename and change the javadoc?

apple502j commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	/**
+	 * Gets the sky property registered for an id. Returns the overworld sky property if
```
Same goes here; maybe rename?

apple502j commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,107 @@
+			throw new IllegalStateException("This id already has a registered skyproperties.");
+		} else {
+			DimensionEffectsAccessor.getIdentifierMap().put(key, properties);
+			System.out.println(DimensionEffectsAccessor.getIdentifierMap().get(key));
```
This should be removed.

apple502j commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,107 @@
+		}
+	}
+
+	public static void registerSkyProperties(Identifier key, DimensionEffects properties) {
```
Rename?

apple502j commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,107 @@
+	}
+
+	@Nullable
+	public static DimensionEffects getSkyProperties(Identifier key) {
```
Rename?

apple502j reviewed (commented):
Renames, etc.

Waterpicker committed:
Pushed suggestions of Haykam and apple502j

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,142 @@
+	 * @param properties The {@link DimensionEffects} option.
+	 * @throws IllegalArgumentException if key is already registered.
+	 */
+	static void registerDimensionEffectss(Identifier key, DimensionEffects properties) {
```
This typo should be fixed, and the parameter + implementation method should refer to "effects" instead of "(sky) properties".

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,142 @@
+	}
+
+	/**
+	 * Registers a dimension effect for a {@link net.minecraft.util.Identifier}.
```
```suggestion
	 * Registers dimension effects for an {@link Identifier}.
```

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,142 @@
+	}
+
+	/**
+	 * Gets the dimension effect registered for an id. Returns the overworld dimension effect if
```
```suggestion
	 * Gets the dimension effects registered for an id. Returns the overworld dimension effect if
```

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,142 @@
+	 */
+	@Nullable
+	static DimensionEffects getDimensionEffects(Identifier key) {
+		return DimensionRenderingRegistryImpl.getSkyProperties(key);
```
This impl method should also refer to dimension effects instead of sky properties.

Juuxel reviewed (changes requested):
The API itself is good, but there are some outdated names and typos that should be fixed.

Waterpicker committed:
Added missing changes from sky properties to dimension effects.

Juuxel commented on fabric-rendering-v1/src/main/resources/fabric-rendering-v1.mixins.json:
```diff
@@ -14,7 +14,8 @@
     "MixinEntityRenderers",
-    "ScreenMixin"
+    "ScreenMixin",
+    "DimensionEffectsAccessor"
```
Minor: I think the convention is to alphabetically sort these, but it doesn't matter imo

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/impl/client/rendering/DimensionRenderingRegistryImpl.java:
```diff
@@ -0,0 +1,106 @@
+public final class DimensionRenderingRegistryImpl {
+	private static final Map<RegistryKey<World>, SkyRenderer> SKY_RENDERERS = new IdentityHashMap<>();
+	private static final Map<RegistryKey<World>, CloudRenderer> CLOUD_RENDERERS = new HashMap<>();
+	private static final Map<RegistryKey<World>, WeatherRenderer> WEATHER_RENDERERS = new HashMap<>();
```
Just out of curiosity, why not all `IdentityHashMap`s or `HashMap`s? iirc registry keys are unique and backed with some weird eternal cache

liach replied:
Yeah, so a hash map is enough to handle all registry keys.

Waterpicker replied:
So do I need to change something?

Juuxel replied:
Not really, this works just fine. It's just a bit odd that the code uses two different maps when either one would suffice.

Waterpicker replied:
Oh I see it now. That's more a byproduct of most likely being tired.

Waterpicker committed:
Corrected me being inconsistent.

liach commented:
@Waterpicker i would recommend using regular hash maps to store registry keys as regular hash maps are more optimized in the jdk.

modmuss50 commented:
The build is failing becuase of checkstyle, please also take a look at review feedback.

Technici4n commented:
> i would recommend using regular hash maps to store registry keys as regular hash maps are more optimized in the jdk.

Quite the opposite I'm pretty sure. Identity hash maps should be faster.

liach commented:
> > i would recommend using regular hash maps to store registry keys as regular hash maps are more optimized in the jdk.
> 
> Quite the opposite I'm pretty sure. Identity hash maps should be faster.

Hmm, do you have a link to one of those demo articles. Last time I read jdk code, identity hash map doesn't handle hash collisions (no treeify) that efficiently it seems

Technici4n commented:
https://stackoverflow.com/questions/17195733/why-identityhashmap-uses-linear-probing-for-collision-resolution

That's my (admittedly not so scientific) reference.

liach commented:
That segment quoted was literally [from more than 14 years ago](https://github.com/openjdk/jdk/blame/30707f80076d2e7a23eb6f6fb1eac2299e64df28/src/java.base/share/classes/java/util/IdentityHashMap.java#L118) (except they moved it into a `@implNote` javadoc tag). Just look at how `HashMap` has custom `computeIfAbsent`, `getOrDefault` implementations (new utilities from Java 8) while `IdentityHashMap` has none of those.

Also a stackoverflow rant:
![stackoverflow rant](https://user-images.githubusercontent.com/7806504/139394670-0fb802aa-cb18-4284-ba5b-273b7b527d6f.png)

Waterpicker committed:
Forgot to remove import.

Technici4n commented:
@liach
![image](https://user-images.githubusercontent.com/13494793/139418244-a6c760ff-3b96-4ecf-9a8d-1abd0b1ca3bb.png)

liach commented:
Well, I should probably check if someone made an jdk suggestion to implement those java 8 map enhancements (which would probably a better way to solve this performance problem)

Waterpicker commented:
> 
> 
> Well, I should probably check if someone made an jdk suggestion to implement those java 8 map enhancements (which would probably a better way to solve this performance problem)

So does this affect my PR in anyway?

modmuss50 commented:
This has been in last call for 7 days, however there are still pending reviews. If possible please take a look them, if you are unable to for what ever reason please let me know and ill happily finish it off and get it merged.

Many thanks 👍

Waterpicker commented:
> 
> 
> This has been in last call for 7 days, however there are still pending reviews. If possible please take a look them, if you are unable to for what ever reason please let me know and ill happily finish it off and get it merged.
> 
> Many thanks 👍

Currently been busy so go ahead.

Waterpicker committed:
Impliments 

https://github.com/FabricMC/fabric/pull/1727#discussion_r734988305
https://github.com/FabricMC/fabric/pull/1727#discussion_r734988265

Waterpicker committed:
Merge remote-tracking branch 'FabricMC/1.17' into environmental-renderers

Waterpicker committed:
Implements https://github.com/FabricMC/fabric/pull/1727#discussion_r730429748

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,143 @@
+	 * <p>This overrides Vanilla's sky rendering.
+	 * @param key A {@link RegistryKey} for your {@link World}
+	 * @param renderer A {@link SkyRenderer} implementation
+	 * @throws IllegalArgumentException if key is already registered.
```
These IAEs aren't thrown anymore so this is outdated (also applies to the other registration methods)

Juuxel commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/DimensionRenderingRegistry.java:
```diff
@@ -0,0 +1,143 @@
+	/**
+	 * Gets the dimension effects registered for an id.
+	 * @param key A {@link RegistryKey} for your {@link World}.
+	 * @return overworld effect if no dimension effects is registered for the key.
```
Doesn't it return null in that case?

Juuxel commented:
(These two issues with the JD can be addressed in a follow-up PR IMO, doesn't really matter that much API-wise. It's just that they're incorrect and the methods don't behave that way.)

Waterpicker committed:
Last minute tweak. Runnable that controls fog runs right before dimension spefific rendering is done. Based on more subtle issues discovered while working api with twilight forest.
